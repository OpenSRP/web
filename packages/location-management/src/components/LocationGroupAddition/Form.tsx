import { Formik } from 'formik';
import * as Yup from 'yup';
import React from 'react';
import { SubmitButton, Form as AntForm, Input, Radio, Select, ResetButton } from 'formik-antd';
import { notification } from 'antd';
import { history } from '@onaio/connected-reducer-registry';

import { KeycloakUser } from '@opensrp/store';
import { OpenSRPService } from '@opensrp/server-service';
import { LocationTagPayloadPOST } from 'location-management/src/ducks/location-tags';

const layout = { labelCol: { span: 8 }, wrapperCol: { span: 8 } };
const offsetLayout = { wrapperCol: { offset: 8, span: 8 } };

/* eslint-disable no-template-curly-in-string */
const validateMessages = {
  required: '${label} is required!',
  number: {
    len: 'lenth should be ${label}',
    min: 'should be greater than ${label}',
    max: 'should be lesser than ${label}',
    range: '${label} must be between ${min} and ${max}',
  },
  default: '${label} defauslt',
  enum: '${label} enum',
  whitespace: '${label} whitespace',
  date: {
    format: '${label} format',
    parse: '${label} parse',
    invalid: '${label} invalid',
  },
  types: {
    string: '${label} is not a valid string',
    method: '${label} is not a valid method',
    array: '${label} is not a valid array',
    object: '${label} is not a valid object',
    number: '${label} is not a valid number',
    date: '${label} is not a valid date',
    boolean: '${label} is not a valid boolean',
    integer: '${label} is not a valid integer',
    float: '${label} is not a valid float',
    regexp: '${label} is not a valid regexp',
    email: '${label} is not a valid email',
    url: '${label} is not a valid url',
    hex: '${label} is not a valid hex',
  },
  string: {
    len: '${label} len',
    min: '${label} min',
    max: '${label} max',
    range: '${label} range',
  },
  array: {
    len: '${label} len',
    min: '${label} min',
    max: '${label} max',
    range: '${label} range',
  },
  pattern: {
    mismatch: '${label} mismatch',
  },
};
/* eslint-enable no-template-curly-in-string */

const location = [
  { name: 'Option 1', value: 'Option1' },
  { name: 'Option 2', value: 'Option2' },
  { name: 'Option 3', value: 'Option3' },
];

const status = [
  { label: 'Active', value: true },
  { label: 'Inactive', value: false },
];

const initialValue = {
  name: '',
  description: '',
  active: 'active',
};

/** yup validations for practitioner data object from form */
export const userSchema = Yup.object().shape({
  name: Yup.string().typeError('Name must be a String').required('Name is Required'),
  active: Yup.boolean().typeError('Status must be a Boolean').required('Status is Required'),
  description: Yup.string()
    .typeError('Description must be a String')
    .required('Description is Required'),
});

interface Props {
  id?: any;
  keycloakUsers: KeycloakUser[];
  accessToken: string;
}

export const Form: React.FC<Props> = (props: Props) => {
  function filter(input: string, option: any) {
    return option.children.toLowerCase().indexOf(input.toLowerCase()) >= 0;
  }

  return (
    <Formik
      initialValues={initialValue}
      validationSchema={userSchema}
      onSubmit={(values, { setSubmitting }) => {
        console.log('values :', JSON.stringify(values));

        const serve = new OpenSRPService(
          props.accessToken,
          'https://opensrp-stage.smartregister.org/opensrp/rest/',
          'location-tag'
        );

        const payload: LocationTagPayloadPOST = { ...values, active: values.active === 'active' };
        serve
          .create(payload)
          .then(() => {
            notification.success({ message: 'User created successfully', description: '' });
            setSubmitting(false);
            history.goBack();
          })
          .catch((e: Error) => {
            notification.error({ message: `${e}`, description: '' });
            setSubmitting(false);
          });
      }}
    >
      {({ errors, isSubmitting, handleSubmit }) => {
        console.log('errors :', errors);

        return (
          <AntForm
            validateMessages={validateMessages}
            requiredMark={false}
            {...layout}
            onSubmitCapture={handleSubmit}
          >
            <AntForm.Item label="Location Name" name="name">
              <Select
                name="name"
                showSearch
                placeholder="Enter a location group name"
                optionFilterProp="children"
                filterOption={filter}
              >
                {location.map((e) => (
                  <Select.Option key={e.value} value={e.value}>
                    {e.name}
                  </Select.Option>
                ))}
              </Select>
            </AntForm.Item>

            <AntForm.Item
              label="Status"
              name="active"
              valuePropName="checked"
              // className={errors.active ? `form-control is-invalid` : `form-control`}
            >
              <Radio.Group name="active" defaultValue={initialValue.active}>
                {status.map((e) => (
                  <Radio name="active" key={e.label} value={e.value}>
                    {e.label}
                  </Radio>
                ))}
              </Radio.Group>
            </AntForm.Item>

            <AntForm.Item
              // className={errors.type ? `form-control is-invalid` : `form-control`}
              name="description"
              label="Type"
            >
              <Input.TextArea name="description" rows={4} placeholder="Description" />
            </AntForm.Item>

            <AntForm.Item name={'buttons'} {...offsetLayout}>
              <SubmitButton>{isSubmitting ? 'Saving' : 'Save'}</SubmitButton>
              <ResetButton onClick={() => history.goBack()}>Cancel</ResetButton>
            </AntForm.Item>
          </AntForm>
        );
      }}
    </Formik>
  );
};

export default Form;
