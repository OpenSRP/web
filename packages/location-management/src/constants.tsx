// URLs
export const URL_ALL_LOCATION_TAGS = '/location-tag';
export const URL_DELETE_LOCATION_TAGS = '/location-tag/delete';
export const KEYCLOAK_API_BASE_URL ='https://opensrp-stage.smartregister.org/opensrp/rest';
