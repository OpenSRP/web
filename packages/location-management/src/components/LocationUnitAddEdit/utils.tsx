import { LocationUnit } from 'location-management/src/ducks/location-units';
import { defaultFormField, FormFields } from './Form';

export enum FormInstances {
  CORE = 'core',
  EUSM = 'eusm',
}

export const getLocationFormFields = (
  location?: LocationUnit,
  instance: FormInstances = FormInstances.CORE
): FormFields => {
  if (!location) {
    return {
      ...defaultFormField,
      instance,
    };
  }
  const formFields = {
    id: location.id,
    instance,
    ...location.properties,
    locationTags: location.locationTags?.map((loc) => loc.id),
    geometry: JSON.stringify(location.geometry),
    type: location.type,
  };

  // remove the existing formFields
  delete formFields.serviceTypes;

  const finalFormFields = {
    ...formFields,
    serviceTypes: location.properties.serviceTypes?.map((type) => type.name) ?? [],
  };

  return finalFormFields;
};
