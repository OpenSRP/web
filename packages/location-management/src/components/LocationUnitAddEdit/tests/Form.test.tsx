/* eslint-disable @typescript-eslint/camelcase */
import flushPromises from 'flush-promises';
import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { store } from '@opensrp/store';
import { notification } from 'antd';
import fetch from 'jest-fetch-mock';
import { formValue, locationUnitgroups, parsedHierarchy, locationSettings } from './fixtures';
import { LocationForm, findParentGeoLocation, onSubmit, removeEmptykeys } from '../Form';
import { act } from 'react-dom/test-utils';
import { history } from '@onaio/connected-reducer-registry';
import { authenticateUser } from '@onaio/session-reducer';
import { baseURL, ERROR_OCCURED } from '../../../constants';
import toJson from 'enzyme-to-json';
import { defaultFormField, FormInstances } from '../utils';

describe('location-management/src/components/LocationUnitAddEdit', () => {
  beforeAll(() => {
    store.dispatch(
      authenticateUser(
        true,
        {
          email: 'bob@example.com',
          name: 'Bobbie',
          username: 'RobertBaratheon',
        },
        // eslint-disable-next-line @typescript-eslint/camelcase
        { api_token: 'hunter2', oAuth2Data: { access_token: 'sometoken', state: 'abcde' } }
      )
    );
  });

  beforeEach(() => {
    fetch.resetMocks();
  });

  it('renders without crashing', async () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm
            openSRPBaseURL={baseURL}
            locationUnitGroup={locationUnitgroups}
            treeData={parsedHierarchy}
          />
        </Router>
      </Provider>
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.find('form')).toHaveLength(1);
  });

  it('creates new location unit', async () => {
    const mockNotificationSuccess = jest.spyOn(notification, 'success');

    await onSubmit(
      jest.fn,
      formValue,
      baseURL,
      locationUnitgroups,
      parsedHierarchy,
      'user_test',
      locationSettings
    );
    await act(async () => {
      await flushPromises();
    });

    expect(fetch.mock.calls[0]).toEqual([
      'https://opensrp-stage.smartregister.org/opensrp/rest/location?is_jurisdiction=true',
      {
        'Cache-Control': 'no-cache',
        Pragma: 'no-cache',
        body: fetch.mock.calls[0][1].body,
        headers: {
          accept: 'application/json',
          authorization: 'Bearer sometoken',
          'content-type': 'application/json;charset=UTF-8',
        },
        method: 'POST',
      },
    ]);

    expect(mockNotificationSuccess).toHaveBeenCalledWith({
      description: undefined,
      message: 'Location Unit Created successfully',
    });
  });

  it('test removeEmptykeys from payload ', async () => {
    const obj = {
      key: 'value',
      testa: undefined,
      testb: [],
      testc: null,
      testd: '',
      child: {
        key: 'value',
        testa: undefined,
        testb: [],
        testc: null,
      },
    };
    removeEmptykeys(obj);

    expect(obj).toMatchObject({
      key: 'value',
      child: { key: 'value' },
    });
  });

  it('checks geo level is calculated correctly', async () => {
    expect(findParentGeoLocation(parsedHierarchy, '400e9d97-4640-44f5-af54-6f4b314384f5')).toEqual(
      5
    );

    expect(findParentGeoLocation(parsedHierarchy, 'a26ca9c8-1441-495a-83b6-bb5df7698996')).toEqual(
      0
    );

    expect(findParentGeoLocation(parsedHierarchy, 'test')).toEqual(undefined);
  });

  it('handles error when creating new location unit', async () => {
    fetch.mockReject(() => Promise.reject(ERROR_OCCURED));
    const mockNotificationError = jest.spyOn(notification, 'error');

    await onSubmit(
      jest.fn,
      formValue,
      baseURL,
      locationUnitgroups,
      parsedHierarchy,
      'user_test',
      []
    );
    await act(async () => {
      await flushPromises();
    });

    expect(mockNotificationError).toHaveBeenCalledWith({
      description: undefined,
      message: ERROR_OCCURED,
    });
  });

  it('edits location unit successfully', async () => {
    const mockNotificationSuccess = jest.spyOn(notification, 'success');
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm
            openSRPBaseURL={baseURL}
            locationUnitGroup={locationUnitgroups}
            treeData={parsedHierarchy}
          />
        </Router>
      </Provider>
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });
    await onSubmit(
      jest.fn,
      {
        ...formValue,
        parentId: '400e9d97-4640-44f5-af54-6f4b314384f5',
        id: '1',
        serviceTypes: ['Test service type'],
      },
      baseURL,
      locationUnitgroups,
      parsedHierarchy,
      'user_test',
      []
    );

    expect(fetch.mock.calls[1]).toEqual([
      'https://opensrp-stage.smartregister.org/opensrp/rest/location?is_jurisdiction=true',
      {
        'Cache-Control': 'no-cache',
        Pragma: 'no-cache',
        body: JSON.stringify({
          is_jurisdiction: true,
          properties: {
            geographicLevel: 6,
            username: 'user_test',
            parentId: '400e9d97-4640-44f5-af54-6f4b314384f5',
            name: 'Tunisia',
            name_en: 'Tunisia',
            status: 'Active',
            serviceTypes: [{ name: 'Test service type' }],
          },
          id: '1',
          syncStatus: 'Synced',
          type: 'Feature',
          locationTags: [{ id: 2, name: 'Sample 2' }],
        }),
        headers: {
          accept: 'application/json',
          authorization: 'Bearer sometoken',
          'content-type': 'application/json;charset=UTF-8',
        },
        method: 'PUT',
      },
    ]);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(mockNotificationSuccess).toHaveBeenCalledWith({
      description: undefined,
      message: 'Location Unit Updated successfully',
    });
  });

  it('fail geoloaction fetch when create location unit', async () => {
    const mockNotificationError = jest.spyOn(notification, 'error');

    await onSubmit(
      jest.fn,
      { ...formValue, parentId: 'test' },
      baseURL,
      locationUnitgroups,
      parsedHierarchy,
      'user_test'
    ).catch((e) => {
      expect(e.message).toEqual(ERROR_OCCURED);
    });

    expect(mockNotificationError).toHaveBeenCalledWith({
      description: undefined,
      message: ERROR_OCCURED,
    });
  });

  it('fail create location unit', async () => {
    fetch.mockReject();

    const mockNotificationError = jest.spyOn(notification, 'error');

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm
            openSRPBaseURL={baseURL}
            locationUnitGroup={locationUnitgroups}
            treeData={parsedHierarchy}
          />
        </Router>
      </Provider>
    );

    await onSubmit(jest.fn, formValue, baseURL, locationUnitgroups, parsedHierarchy, 'user_test');

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(fetch.mock.calls[1]).toEqual([
      'https://opensrp-stage.smartregister.org/opensrp/rest/location?is_jurisdiction=true',
      {
        'Cache-Control': 'no-cache',
        Pragma: 'no-cache',
        body: JSON.stringify({
          is_jurisdiction: true,
          properties: {
            geographicLevel: 0,
            username: 'user_test',
            name: 'Tunisia',
            name_en: 'Tunisia',
            status: 'Active',
          },
          id: JSON.parse(fetch.mock.calls[1][1].body as string).id,
          syncStatus: 'Synced',
          type: 'Feature',
          locationTags: [{ id: 2, name: 'Sample 2' }],
        }),
        headers: {
          accept: 'application/json',
          authorization: 'Bearer sometoken',
          'content-type': 'application/json;charset=UTF-8',
        },
        method: 'POST',
      },
    ]);

    expect(mockNotificationError).toHaveBeenCalledWith({
      description: undefined,
      message: ERROR_OCCURED,
    });
  });

  it('handles error when editing location unit', async () => {
    fetch.mockReject(() => Promise.reject(ERROR_OCCURED));
    const mockNotificationError = jest.spyOn(notification, 'error');

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm
            openSRPBaseURL={baseURL}
            locationUnitGroup={locationUnitgroups}
            treeData={parsedHierarchy}
          />
        </Router>
      </Provider>
    );
    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(mockNotificationError).toHaveBeenCalledWith({
      description: undefined,
      message: ERROR_OCCURED,
    });
  });

  it('handle failed on submit', async () => {
    fetch.mockReject(() => Promise.reject(ERROR_OCCURED));
    const mockNotificationError = jest.spyOn(notification, 'error');
    await onSubmit(
      jest.fn,
      formValue,
      baseURL,
      locationUnitgroups,
      parsedHierarchy,
      'user_test',
      []
    );

    await act(async () => {
      await flushPromises();
    });

    expect(mockNotificationError).toHaveBeenCalledWith({
      description: undefined,
      message: ERROR_OCCURED,
    });
  });

  it('renders without crashing with id', async () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm
            openSRPBaseURL={baseURL}
            initialValues={formValue}
            locationUnitGroup={locationUnitgroups}
            treeData={parsedHierarchy}
          />
        </Router>
      </Provider>
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.find('form input[name="name"]').prop('value')).toBe(formValue.name);

    expect(
      wrapper.find(`form input[type="radio"][value="${formValue.status}"]`).prop('checked')
    ).toBe(true);

    expect(wrapper.find('form TreeSelect[className="ant-tree-select"]').prop('value')).toBe(
      formValue.parentId
    );
    expect(
      wrapper.find('form Field[name="locationTags"] Select[prefixCls="ant-select"]').prop('value')
    ).toBe(formValue.locationTags);
    expect(wrapper.find('form input[name="type"]').prop('value')).toBe(formValue.type);
  });

  it('Cancel button', async () => {
    const mockBack = jest.fn();
    history.goBack = mockBack;

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm
            openSRPBaseURL={baseURL}
            locationUnitGroup={locationUnitgroups}
            treeData={parsedHierarchy}
          />
        </Router>
      </Provider>
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    wrapper.find('button#cancel').simulate('click');

    // click go back
    expect(wrapper.find('button').first().text()).toMatchInlineSnapshot(`"Save"`);
    wrapper.find('button').first().simulate('click');

    // click go back
    expect(wrapper.find('button').last().text()).toMatchInlineSnapshot(`"Cancel"`);
    wrapper.find('button').last().simulate('click');

    expect(mockBack).toHaveBeenCalled();
  });

  // TODO - review: what does this test check? where is the assertion
  it('Update LocationUnitGroupValue', async () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm
            openSRPBaseURL={baseURL}
            initialValues={formValue}
            locationUnitGroup={locationUnitgroups}
            treeData={parsedHierarchy}
          />
        </Router>
      </Provider>
    );

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await flushPromises();
    });
  });

  // TODO - review: what does this test check? where is the assertion
  it('Create LocationUnitGroupValue', async () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm
            openSRPBaseURL={baseURL}
            initialValues={formValue}
            locationUnitGroup={locationUnitgroups}
            treeData={parsedHierarchy}
          />
        </Router>
      </Provider>
    );

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await flushPromises();
    });
  });
});

describe('LocationForm', () => {
  afterEach(() => {
    fetch.resetMocks();
  });
  it('renders correctly', () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm />
        </Router>
      </Provider>,
      { attachTo: div }
    );

    expect(toJson(wrapper.find('#instance label'))).toMatchSnapshot('instance label');
    expect(toJson(wrapper.find('#instance input'))).toMatchSnapshot('instance field');

    expect(toJson(wrapper.find('#parentId label'))).toMatchSnapshot('parentId label');
    expect(toJson(wrapper.find('#parentId select'))).toMatchSnapshot('parentId field');

    expect(toJson(wrapper.find('#name label'))).toMatchSnapshot('name label');
    expect(toJson(wrapper.find('#name input'))).toMatchSnapshot('name field');

    expect(toJson(wrapper.find('#status label').first())).toMatchSnapshot('status label');
    expect(toJson(wrapper.find('#status input'))).toMatchSnapshot('status field');

    expect(toJson(wrapper.find('#isJurisdiction label').first())).toMatchSnapshot(
      'isJurisdiction label'
    );
    expect(toJson(wrapper.find('#isJurisdiction input'))).toMatchSnapshot('isJurisdiction field');
    expect(toJson(wrapper.find('#type label'))).toMatchSnapshot('type label');
    expect(toJson(wrapper.find('#type input'))).toMatchSnapshot('type field');
    expect(toJson(wrapper.find('#serviceTypes label'))).toMatchSnapshot('serviceTypes label');
    expect(toJson(wrapper.find('#serviceTypes input'))).toMatchSnapshot('serviceTypes field');
    expect(toJson(wrapper.find('#externalId label'))).toMatchSnapshot('externalId label');
    expect(toJson(wrapper.find('#externalId input'))).toMatchSnapshot('externalId field');
    expect(toJson(wrapper.find('#geometry label'))).toMatchSnapshot('geometry label');
    expect(toJson(wrapper.find('#geometry input'))).toMatchSnapshot('geometry field');
    expect(toJson(wrapper.find('#locationTags label'))).toMatchSnapshot('locationTags label');
    expect(toJson(wrapper.find('#locationTags input'))).toMatchSnapshot('locationTags field');

    wrapper.find('.extraFields').forEach((field) => {
      expect(toJson(field.find('label'))).toMatchSnapshot('field label');
      expect(toJson(field.find('input'))).toMatchSnapshot('field input');
    });

    expect(toJson(wrapper.find('#submit button'))).toMatchSnapshot('submit button');
    expect(toJson(wrapper.find('#cancel button'))).toMatchSnapshot('cancel button');
  });

  it('form validation works for core instance', async () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    // when instance is set to core by default, types is required, serviceTypes is not required
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm />
        </Router>
      </Provider>,
      { attachTo: div }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.find('FormItem#instance').text()).toMatchInlineSnapshot(`"Instance"`);

    expect(wrapper.find('FormItem#parentId').text()).toMatchInlineSnapshot(`"ParentPlease select"`);

    expect(wrapper.find('FormItem#status').text()).toMatchInlineSnapshot(`"StatusActiveInactive"`);

    expect(wrapper.find('FormItem#isJurisdiction').text()).toMatchInlineSnapshot(
      `"Location Category(optional)Service pointJurisdiction"`
    );

    // type is required for core
    expect(wrapper.find('FormItem#type').text()).toMatchInlineSnapshot(`"TypeType is Required"`);

    // not required for core
    expect(wrapper.find('FormItem#serviceTypes').text()).toMatchInlineSnapshot(`"Type"`);

    expect(wrapper.find('FormItem#externalId').text()).toMatchInlineSnapshot(
      `"External ID(optional)"`
    );

    expect(wrapper.find('FormItem#geometry').text()).toMatchInlineSnapshot(`"geometry(optional)"`);

    expect(wrapper.find('FormItem#locationTags').text()).toMatchSnapshot(
      'location Tags does not have error message'
    );

    wrapper.unmount();
  });

  it('form validation works for eusm instance', async () => {
    const div = document.createElement('div');
    document.body.appendChild(div);

    // when instance is set to core by default, types is required, serviceTypes is not required
    const initialValues = { ...defaultFormField, instance: FormInstances.EUSM };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <LocationForm initialValues={initialValues} />
        </Router>
      </Provider>,
      {
        attachTo: div,
      }
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.find('FormItem#instance').text()).toMatchInlineSnapshot(`"Instance"`);

    expect(wrapper.find('FormItem#parentId').text()).toMatchInlineSnapshot(`"ParentPlease select"`);

    expect(wrapper.find('FormItem#status').text()).toMatchInlineSnapshot(`"StatusActiveInactive"`);

    expect(wrapper.find('FormItem#isJurisdiction').text()).toMatchInlineSnapshot(
      `"Location Category(optional)Service pointJurisdiction"`
    );

    // type is required for core, but not required for eusm instance
    expect(wrapper.find('FormItem#type').text()).toMatchInlineSnapshot(`"Type"`);

    // service types is required for eusm
    expect(wrapper.find('FormItem#serviceTypes').text()).toMatchInlineSnapshot(
      `"TypeService Types is required"`
    );

    expect(wrapper.find('FormItem#externalId').text()).toMatchInlineSnapshot(
      `"External ID(optional)"`
    );

    expect(wrapper.find('FormItem#geometry').text()).toMatchInlineSnapshot(`"geometry(optional)"`);

    expect(wrapper.find('FormItem#locationTags').text()).toMatchSnapshot(
      'location Tags does not have error message'
    );
    wrapper.unmount();
  });
});
