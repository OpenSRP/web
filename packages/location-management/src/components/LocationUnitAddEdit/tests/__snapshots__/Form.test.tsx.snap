// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationForm form validation works for core instance: location Tags does not have error message 1`] = `"Unit Group(optional) Enter a location group name"`;

exports[`LocationForm form validation works for eusm instance: location Tags does not have error message 1`] = `"Unit Group(optional) Enter a location group name"`;

exports[`LocationForm renders correctly: cancel button 1`] = `
<button
  className="ant-btn ant-btn-dashed"
  id="cancel"
  onClick={[Function]}
  type="button"
>
  <LoadingIcon
    existIcon={false}
    loading={false}
    prefixCls="ant-btn"
  >
    <CSSMotion
      motionName="ant-btn-loading-icon-motion"
      onAppearActive={[Function]}
      onAppearStart={[Function]}
      onEnterActive={[Function]}
      onEnterStart={[Function]}
      onLeaveActive={[Function]}
      onLeaveStart={[Function]}
      removeOnLeave={true}
      visible={false}
    >
      <DomWrapper />
    </CSSMotion>
  </LoadingIcon>
  <span
    key=".0"
  >
    Cancel
  </span>
</button>
`;

exports[`LocationForm renders correctly: externalId field 1`] = `
<input
  className="ant-input"
  name="externalId"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  placeholder="Select status"
  type="text"
  value=""
/>
`;

exports[`LocationForm renders correctly: externalId label 1`] = `
<label
  className="ant-form-item-required-mark-optional"
  title="External ID"
>
  External ID
  <span
    className="ant-form-item-optional"
  >
    (optional)
  </span>
</label>
`;

exports[`LocationForm renders correctly: geometry field 1`] = `null`;

exports[`LocationForm renders correctly: geometry label 1`] = `
<label
  className="ant-form-item-required-mark-optional"
  title="geometry"
>
  geometry
  <span
    className="ant-form-item-optional"
  >
    (optional)
  </span>
</label>
`;

exports[`LocationForm renders correctly: instance field 1`] = `
<input
  className="ant-input"
  name="instance"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  type="text"
  value="core"
/>
`;

exports[`LocationForm renders correctly: instance label 1`] = `
<label
  className="ant-form-item-required ant-form-item-required-mark-optional"
  title="Instance"
>
  Instance
</label>
`;

exports[`LocationForm renders correctly: isJurisdiction field 1`] = `
Array [
  <input
    checked={true}
    className="ant-radio-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value={false}
  />,
  <input
    checked={false}
    className="ant-radio-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value={true}
  />,
]
`;

exports[`LocationForm renders correctly: isJurisdiction label 1`] = `
<label
  className="ant-form-item-required-mark-optional"
  title="Location Category"
>
  Location Category
  <span
    className="ant-form-item-optional"
  >
    (optional)
  </span>
</label>
`;

exports[`LocationForm renders correctly: locationTags field 1`] = `
<input
  aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
  aria-autocomplete="list"
  aria-controls="rc_select_TEST_OR_SSR_list"
  aria-haspopup="listbox"
  aria-owns="rc_select_TEST_OR_SSR_list"
  autoComplete="off"
  className="ant-select-selection-search-input"
  id="rc_select_TEST_OR_SSR"
  onChange={[Function]}
  onCompositionEnd={[Function]}
  onCompositionStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onPaste={[Function]}
  readOnly={true}
  role="combobox"
  style={
    Object {
      "opacity": 0,
    }
  }
  type="search"
  unselectable="on"
  value=""
/>
`;

exports[`LocationForm renders correctly: locationTags label 1`] = `
<label
  className="ant-form-item-required-mark-optional"
  title="Unit Group"
>
  Unit Group
  <span
    className="ant-form-item-optional"
  >
    (optional)
  </span>
</label>
`;

exports[`LocationForm renders correctly: name field 1`] = `
<input
  className="ant-input"
  name="name"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  placeholder="Enter a location name"
  type="text"
  value=""
/>
`;

exports[`LocationForm renders correctly: name label 1`] = `
<label
  className="ant-form-item-required ant-form-item-required-mark-optional"
  title="Name"
>
  Name
</label>
`;

exports[`LocationForm renders correctly: parentId field 1`] = `null`;

exports[`LocationForm renders correctly: parentId label 1`] = `
<label
  className="ant-form-item-required ant-form-item-required-mark-optional"
  title="Parent"
>
  Parent
</label>
`;

exports[`LocationForm renders correctly: serviceTypes field 1`] = `
<input
  aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
  aria-autocomplete="list"
  aria-controls="rc_select_TEST_OR_SSR_list"
  aria-haspopup="listbox"
  aria-owns="rc_select_TEST_OR_SSR_list"
  autoComplete="off"
  className="ant-select-selection-search-input"
  id="rc_select_TEST_OR_SSR"
  onChange={[Function]}
  onCompositionEnd={[Function]}
  onCompositionStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onPaste={[Function]}
  readOnly={true}
  role="combobox"
  style={
    Object {
      "opacity": 0,
    }
  }
  type="search"
  unselectable="on"
  value=""
/>
`;

exports[`LocationForm renders correctly: serviceTypes label 1`] = `
<label
  className="ant-form-item-required ant-form-item-required-mark-optional"
  title="Type"
>
  Type
</label>
`;

exports[`LocationForm renders correctly: status field 1`] = `
Array [
  <input
    checked={true}
    className="ant-radio-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value="Active"
  />,
  <input
    checked={false}
    className="ant-radio-input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value="InActive"
  />,
]
`;

exports[`LocationForm renders correctly: status label 1`] = `
<label
  className="ant-form-item-required ant-form-item-required-mark-optional"
  title="Status"
>
  Status
</label>
`;

exports[`LocationForm renders correctly: submit button 1`] = `
<button
  className="ant-btn ant-btn-primary"
  disabled={false}
  id="submit"
  onClick={[Function]}
  type="submit"
>
  <LoadingIcon
    existIcon={false}
    loading={false}
    prefixCls="ant-btn"
  >
    <CSSMotion
      motionName="ant-btn-loading-icon-motion"
      onAppearActive={[Function]}
      onAppearStart={[Function]}
      onEnterActive={[Function]}
      onEnterStart={[Function]}
      onLeaveActive={[Function]}
      onLeaveStart={[Function]}
      removeOnLeave={true}
      visible={false}
    >
      <DomWrapper />
    </CSSMotion>
  </LoadingIcon>
  <span
    key=".0"
  >
    Save
  </span>
</button>
`;

exports[`LocationForm renders correctly: type field 1`] = `
<input
  className="ant-input"
  name="type"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  placeholder="Select type"
  type="text"
  value=""
/>
`;

exports[`LocationForm renders correctly: type label 1`] = `
<label
  className="ant-form-item-required ant-form-item-required-mark-optional"
  title="Type"
>
  Type
</label>
`;
